lib/compat/wordpress-5.9/class-gutenberg-rest-templates-controller.php

```diff
@@ -332,6 +332,61 @@ public function delete_item( $request ) {
 		return $this->prepare_item_for_response( $template, $request );
 	}
 
+	/**
+	 * Requesting this endpoint for a template like "twentytwentytwo//home" requires using
+	 * a path like /wp/v2/templates/twentytwentytwo//home. There are special cases when
+	 * WordPress routing corrects the name to contain only a single slash like "twentytwentytwo/home".
+	 *
+	 * This method attempts to find a template with a specific ID, and if it's missing then it
+	 * falls back to parsing REQUEST_URI in an attempt to grab the verbatim ID passed by the user.
+	 *
+	 * See https://core.trac.wordpress.org/ticket/54507 for more context
+	 *
+	 * @param string  $id        ID of the template.
+	 * @param boolean $from_file Optional Whether to use gutenberg_get_block_template instead of get_block_file_template.
+	 * @return WP_Block_Template|null Template.
+	 */
+	private function get_template_by_id( $id, $from_file = false ) {
+		$getter   = $from_file ? 'get_block_file_template' : 'gutenberg_get_block_template';
+		$template = $getter( $id, $this->post_type );
+		if ( ! $template ) {
+			// Make sure REQUEST_URI is set.
+			if ( empty( $_SERVER['REQUEST_URI'] ) ) {
+				return;
+			}
+
+			$uri       = $_SERVER['REQUEST_URI'];
+			$delimiter = $this->namespace . '/' . $this->rest_base . '/';
+
+			// Extract the part of the path that comes after /wp/v2/templates/.
+			$start = strpos( $uri, $delimiter );
+			if ( false === $start ) {
+				return;
+			}
+			$inferred_id = substr( $uri, $start + strlen( $delimiter ) );
+
+			// Only use part of the path until the first querystring-like symbol (either ? or &).
+			$parts = preg_split( '|[?&]|', $inferred_id );
```

Comment:
Similarly, should we be stricter and match any character disallowed by the ID format? For example, anything other than `[A-z0-9_-]`?

Reply:
I actually played with that initially. This would work under the assumption that template ID is always generated like `$theme//$slug` and that both `$theme` and `$slug` adhere to this format. I think that will be true in most cases so hm yes, let's try it.

But it also gave me another idea. Given that `$slug` can't contain a slash, we only have to try doubling the last slash in `$request['id']` – it should be much simpler than all this parsing. Let's try it.